# Домашняя работа по теме "Машинное обучение ранжированию"

В этом ДЗ мы:
- научимся работать со стандартным датасетом для машинного обучения ранжированию [MSLR](https://www.microsoft.com/en-us/research/project/mslr/)
- попробуем применить на практике все то, чему мы научились на семинаре

## Как будет происходить сдача ДЗ

Вам надо:
- форкнуть эту репу
- создать бранч в котором вы дальше будете работать
- реализовать класс Model в этом ноутбуке
- убедиться, что ваша реализация выбивает NDCG@10 выше бейзлайна (см. ниже)
- запушить ваш бранч и поставить Pull Request
- в комментарии написать какой скор вы выбили

В таком случае мы (организаторы):

- счекаутим вашу бранчу
- проверим что ваша реализация действительно выбивает заявленный скор

Предполагается, что и вы, и мы работаем в виртаульном окружении как в семинаре про машинное обучение ранжированию: seminars/7-learning-to-rank/requirements.txt(подробнее про работу с виртуальными окружениями README в корне этой репы).

Оценка:
- За выбитый скор больше **0.507** назначаем **5** баллов, за скор больше (или равно) **0.510** назначаем максимальный балл -- 10 баллов
- Тот из участников кто выбъет самый высокий скор получит еще +10 баллов

При сдаче кода важно помнить о том, что:
- В коде не должно быть захардкоженных с потолка взятых гиперпараметров (таких как число деревьев, learning rate и т.п.) -- обязательно должен быть представлен код который их подбирает!
- Решение должно быть стабильно от запуска к запуску (на CPU) т.е. все seed'ы для генераторов случайных чисел должны быть фиксированы
- Мы (организаторы) будем запускать код на CPU поэтому, даже если вы использовали для подбора параметров GPU, финальный скор надо репортить на CPU

## Пререквизиты

Импортируем все что нам понадобится для дальнейшей работы:

In [1]:
import pathlib
from timeit import default_timer as timer

import numpy as np
import pandas as pd

from catboost import datasets, utils

## Датасет MSLR (Microsoft Learning to Rank)

Загрузим датасет MSLR.

Полный датасет можно скачать с официального сайта: https://www.microsoft.com/en-us/research/project/mslr/

Строго говоря, он состоит их 2х частей:

- основной датасет MSLR-WEB30K -- он содержит более 30 тыс. запросов
- "маленький" датасет MSLR-WEB10K, который содержит только 10 тыс. запросов и является случайным сэмплом датасета MSLR-WEB30K

в этом ДЗ мы будем работать с MSLR-WEB10K, т.к. полная версия датасета может просто не поместиться у нас в RAM (и, тем более, в память видеокарты если мы учимся на GPU)

Будем считать, что мы самостоятельно скачали датасет MSLR-WEB10K с официального сайта, поместили его в папку КОРЕНЬ-ЭТОЙ-РЕПЫ/data/mslr-web10k и раззиповали.

В результате у нас должна получиться следующая структура папок:

In [2]:
!ls ../../data

mslr-web10k  vkmarcoranking-v1-document


In [3]:
!ls -lh ../../data/mslr-web10k/

total 20K
drwxrwxr-x 2 jovyan jovyan 4.0K Apr 20 16:04 Fold1
drwxrwxr-x 2 jovyan jovyan 4.0K Apr 20 16:04 Fold2
drwxrwxr-x 2 jovyan jovyan 4.0K Apr 20 16:03 Fold3
drwxrwxr-x 2 jovyan jovyan 4.0K Apr 20 16:03 Fold4
drwxrwxr-x 2 jovyan jovyan 4.0K Apr 20 16:04 Fold5


Заметим, что датасет довольно большой, в распакованном виде он весит 7.7 GB.

Датасет состоит из нескольких фолдов, которые по сути представляют из себя разные разбиения одних и тех же данных на обучающее, валидационное и тестовые множеста.

Дальше мы будем использовать только первый фолд: Fold1.

Заглянем внутрь:

In [4]:
!ls -lh ../../data/mslr-web10k/Fold1

total 1.3G
-rw-r--r-- 1 jovyan jovyan 267M Apr 20 16:04 test.txt
-rw-r--r-- 1 jovyan jovyan 800M Apr 20 16:04 train.txt
-rw-r--r-- 1 jovyan jovyan 261M Apr 20 16:04 vali.txt


Видим, что у нас 3 файла с говорящими названиями, соответсвующими сплитам нашего датасета.

Посмотрим на содержимое одного из файлов:

In [5]:
!head -n 1 ../../data/mslr-web10k/Fold1/train.txt

2 qid:1 1:3 2:3 3:0 4:0 5:3 6:1 7:1 8:0 9:0 10:1 11:156 12:4 13:0 14:7 15:167 16:6.931275 17:22.076928 18:19.673353 19:22.255383 20:6.926551 21:3 22:3 23:0 24:0 25:6 26:1 27:1 28:0 29:0 30:2 31:1 32:1 33:0 34:0 35:2 36:1 37:1 38:0 39:0 40:2 41:0 42:0 43:0 44:0 45:0 46:0.019231 47:0.75000 48:0 49:0 50:0.035928 51:0.00641 52:0.25000 53:0 54:0 55:0.011976 56:0.00641 57:0.25000 58:0 59:0 60:0.011976 61:0.00641 62:0.25000 63:0 64:0 65:0.011976 66:0 67:0 68:0 69:0 70:0 71:6.931275 72:22.076928 73:0 74:0 75:13.853103 76:1.152128 77:5.99246 78:0 79:0 80:2.297197 81:3.078917 82:8.517343 83:0 84:0 85:6.156595 86:2.310425 87:7.358976 88:0 89:0 90:4.617701 91:0.694726 92:1.084169 93:0 94:0 95:2.78795 96:1 97:1 98:0 99:0 100:1 101:1 102:1 103:0 104:0 105:1 106:12.941469 107:20.59276 108:0 109:0 110:16.766961 111:-18.567793 112:-7.760072 113:-20.838749 114:-25.436074 115:-14.518523 116:-21.710022 117:-21.339609 118:-24.497864 119:-27.690319 120:-20.203779 121:-15.449379 122:-4.474452 123:-23.634899 

Видим, что данные лежат в уже знакомом нам по семинару формате:

- В первой колонке лежит таргет (оценка асессора), по 5-балльной шкале релевантности: от 0 до 4 (включительно)
- Во второй колонке лежит ID запроса, по которому можно сгруппировать все оценки документов в рамках одного и того же запроса
- Дальше идет вектор из 128 фичей (таких как значения BM25 и т.п.), их точная природа нам сейчас на важна

В файле qid и все-фичи кодируются в формате КЛЮЧ:ЗНАЧЕНИЕ, напр. 130:116 -- тут 130 это номер фичи, а 116 -- ее значение.

Такой формат в мире машинного обучения часто называют svm light формат (в честь когда-то популярной библиотеки SVM-Light)

Напишем немного вспомогательного кода для загрузки этого датасета:

In [6]:
def generate_column_names(num_features):
    """Generates column names for LETOR-like datasets"""
    columns = ['label', 'qid']
    for i in range(num_features):
        column = f"feature_{i+1}"
        columns.append(column)
    return columns
    
def load_svmlight_file(input_file, max_num_lines=0):
    """Loads dataset split in SVM-Light format"""
    def _parse_field(field):
        parts = field.split(':')
        if len(parts) != 2:
            raise Exception(f"invalid number of parts in field {field}")
        return parts

    num_features = 136
    exp_num_fields = num_features + 2
    num_lines = 0
    X = []
    with open(input_file, 'rt') as f:
        for line in f:
            try:
                num_lines += 1
                                  
                # Parse into fields
                fields = line.rstrip().split(' ')
                num_fields = len(fields)
                if num_fields != exp_num_fields:
                    raise Exception(f"invalid number of fields {num_fields}")
    
                # Parse every field
                x = np.zeros(exp_num_fields, dtype=np.float32)
                label = int(fields[0])
                x[0] = label
                _, qid_str = _parse_field(fields[1])
                qid = int(qid_str)
                x[1] = qid
                for i, field in enumerate(fields[2:]):
                    _, feature_str = _parse_field(field)
                    x[i+2] = float(feature_str)
    
                # Add new object
                X.append(x)
                if num_lines % 50000 == 0:
                    print(f"Loaded {num_lines} lines...")
                if max_num_lines > 0 and num_lines == max_num_lines:
                    print(f"WARNING: stop loading, line limit reached: max_num_lines = {max_num_lines} input_file = {input_file}")
                    break
            except Exception as e:
                raise Exception(f"error at line {num_lines} in {input_file}") from e
    
    # To pandas
    df = pd.DataFrame(X, columns=generate_column_names(num_features))
    print(f"Loaded SVM-Light file {input_file}")
    return df

И теперь загрузим датасет:

In [ ]:
fold_dir = pathlib.Path("../../data/mslr-web10k/Fold1")

df_train = load_svmlight_file(fold_dir.joinpath("train.txt"))
df_valid = load_svmlight_file(fold_dir.joinpath("vali.txt"))
df_test = load_svmlight_file(fold_dir.joinpath("test.txt"))
print(f"Dataset loaded from fold_dir {fold_dir}")

Loaded 50000 lines...
Loaded 100000 lines...
Loaded 150000 lines...
Loaded 200000 lines...
Loaded 250000 lines...
Loaded 300000 lines...
Loaded 350000 lines...
Loaded 400000 lines...
Loaded 450000 lines...
Loaded 500000 lines...
Loaded 550000 lines...
Loaded 600000 lines...
Loaded 650000 lines...
Loaded 700000 lines...
Loaded SVM-Light file ../../data/mslr-web10k/Fold1/train.txt
Loaded 50000 lines...
Loaded 100000 lines...
Loaded 150000 lines...
Loaded 200000 lines...
Loaded SVM-Light file ../../data/mslr-web10k/Fold1/vali.txt
Loaded 50000 lines...
Loaded 100000 lines...
Loaded 150000 lines...
Loaded 200000 lines...
Loaded SVM-Light file ../../data/mslr-web10k/Fold1/test.txt
Dataset loaded from fold_dir ../../data/mslr-web10k/Fold1


Посмотрим на данные:

In [ ]:
print(df_train.head(5))

   label  qid  feature_1  feature_2  feature_3  feature_4  feature_5  \
0    2.0  1.0        3.0        3.0        0.0        0.0        3.0   
1    2.0  1.0        3.0        0.0        3.0        0.0        3.0   
2    0.0  1.0        3.0        0.0        2.0        0.0        3.0   
3    2.0  1.0        3.0        0.0        3.0        0.0        3.0   
4    1.0  1.0        3.0        0.0        3.0        0.0        3.0   

   feature_6  feature_7  feature_8  ...  feature_127  feature_128  \
0        1.0        1.0   0.000000  ...         62.0   11089534.0   
1        1.0        0.0   1.000000  ...         54.0   11089534.0   
2        1.0        0.0   0.666667  ...         45.0          3.0   
3        1.0        0.0   1.000000  ...         56.0   11089534.0   
4        1.0        0.0   1.000000  ...         64.0          5.0   

   feature_129  feature_130  feature_131  feature_132  feature_133  \
0          2.0        116.0      64034.0         13.0          3.0   
1          2

Т.е. теперь мы видим что данные доступны в точно таком же виде, как это было в семинаре.

Проведем небольшой EDA.

Всего у нас 723 тыс. документов в трейне:

In [ ]:
print(df_train.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 723412 entries, 0 to 723411
Columns: 138 entries, label to feature_136
dtypes: float32(138)
memory usage: 380.8 MB
None


235 тыс. документов в валидации:

In [ ]:
print(df_valid.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235259 entries, 0 to 235258
Columns: 138 entries, label to feature_136
dtypes: float32(138)
memory usage: 123.8 MB
None


И 241 тыс. документов в тесте:

In [ ]:
print(df_test.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 241521 entries, 0 to 241520
Columns: 138 entries, label to feature_136
dtypes: float32(138)
memory usage: 127.1 MB
None


Сколько у нас запросов?

In [12]:
num_queries_train = df_train['qid'].nunique()
num_queries_valid = df_valid['qid'].nunique()
num_queries_test = df_test['qid'].nunique()
print(f"Got {num_queries_train} train, {num_queries_valid} valid and {num_queries_test} test queries")

Got 6000 train, 2000 valid and 2000 test queries


## Обучаем модель

Теперь можно приступить непосредственно к обучению модели. 

Объявим класс модели, который надо будем заимлементить в этом ДЗ:

In [23]:
import itertools
from catboost import CatBoostRanker, Pool


class Model:
    def __init__(self, task_type='CPU'):
        self.task_type = task_type
        self.best_score = 0.0  # Инициализируем с бесконечностью, так как ищем минимум
        self.best_params = {}

    @staticmethod
    def to_catboost_dataset(df):
        y = df['label'].to_numpy()                       # Label: [0-4]
        q = df['qid'].to_numpy().astype('uint32')        # Query Id
        X = df.drop(columns=['label', 'qid']).to_numpy() # 136 features
        return (X, y, q)

    def fit(self, df_train, df_valid, params=None):
        X_train, y_train, q_train = self.to_catboost_dataset(df_train)
        train_pool = Pool(data=X_train, label=y_train, group_id=q_train)

        X_valid, y_valid, q_valid = self.to_catboost_dataset(df_valid)
        valid_pool = Pool(data=X_valid, label=y_valid, group_id=q_valid)
        if params is None:
            param_grid = {
                'loss': ['YetiRank', 'QueryRMSE'],
                'learning_rate': [0.03, 0.07],
                'max_bin': [128, 254],
                'depth': [6, 10, 12],
            }
            all_combinations = list(itertools.product(*list(param_grid.values())))
    
            for combination in all_combinations:
                params = {
                    'loss_function': combination[0],
                    'random_seed': 42,
                    'iterations': 8000,
                    'learning_rate': combination[1],
                    'max_bin': combination[2],
                    'depth': combination[3],
                    'task_type': self.task_type,
                    'eval_metric': "NDCG:top=10;type=Exp"
                    
                }
                self.model = CatBoostRanker(**params)
                self.model.fit(train_pool, eval_set=valid_pool, verbose=100, use_best_model=True)
                if np.max(self.model.get_evals_result()["validation"]["NDCG:top=10;type=Exp"]) > self.best_score:
                    self.best_score = np.max(self.model.get_evals_result()["validation"]["NDCG:top=10;type=Exp"])
                    self.best_params = params
                print(f"Current best params: {self.best_params}")
        else:
            self.best_params = params

        self.model = CatBoostRanker(**self.best_params)
        self.model.fit(train_pool, eval_set=valid_pool, verbose=100, use_best_model=True)

    def predict(self, df_test):
        X_test, y_test, q_test = self.to_catboost_dataset(df_test)
        test_pool = Pool(data=X_test, label=y_test, group_id=q_test)
        predictions = self.model.predict(test_pool)
        return predictions


Создадим и применим модель:

In [ ]:
# Create model
model = Model(task_type="GPU")

# Fit
start = timer()
model.fit(df_train, df_valid)
elapsed = timer() - start
print(f"Model fit: elapsed = {elapsed:.3f}")

# Predict
y_hat_test = model.predict(df_test)
print(f"Predicted: y_hat_test.shape = {y_hat_test.shape}")

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2832659	best: 0.2832659 (0)	total: 128ms	remaining: 17m 3s
100:	test: 0.4276595	best: 0.4276595 (100)	total: 1.69s	remaining: 2m 12s
200:	test: 0.4504882	best: 0.4504882 (200)	total: 3.25s	remaining: 2m 6s
300:	test: 0.4599742	best: 0.4599742 (300)	total: 4.8s	remaining: 2m 2s
400:	test: 0.4661336	best: 0.4661336 (400)	total: 6.36s	remaining: 2m
500:	test: 0.4730890	best: 0.4730890 (500)	total: 7.93s	remaining: 1m 58s
600:	test: 0.4785313	best: 0.4785313 (600)	total: 9.49s	remaining: 1m 56s
700:	test: 0.4818239	best: 0.4818239 (700)	total: 11s	remaining: 1m 55s
800:	test: 0.4858823	best: 0.4858823 (800)	total: 12.6s	remaining: 1m 53s
900:	test: 0.4877814	best: 0.4877814 (900)	total: 14.2s	remaining: 1m 51s
1000:	test: 0.4893219	best: 0.4893219 (1000)	total: 15.7s	remaining: 1m 50s
1100:	test: 0.4911008	best: 0.4911008 (1100)	total: 17.3s	remaining: 1m 48s
1200:	test: 0.4930421	best: 0.4930522 (1195)	total: 18.9s	remaining: 1m 46s
1300:	test: 0.4941325	best: 0.4941630 (1275)	

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3454527	best: 0.3454527 (0)	total: 103ms	remaining: 13m 44s
100:	test: 0.4546532	best: 0.4546532 (100)	total: 6.9s	remaining: 8m 59s
200:	test: 0.4690362	best: 0.4690362 (200)	total: 13.8s	remaining: 8m 54s
300:	test: 0.4802397	best: 0.4802397 (300)	total: 20.6s	remaining: 8m 47s
400:	test: 0.4863021	best: 0.4863587 (395)	total: 27.5s	remaining: 8m 41s
500:	test: 0.4925901	best: 0.4925901 (500)	total: 34.4s	remaining: 8m 34s
600:	test: 0.4954280	best: 0.4955442 (595)	total: 41.2s	remaining: 8m 27s
700:	test: 0.4981737	best: 0.4981737 (700)	total: 48.1s	remaining: 8m 20s
800:	test: 0.4992507	best: 0.4992933 (795)	total: 55s	remaining: 8m 14s
900:	test: 0.5009591	best: 0.5010058 (885)	total: 1m 1s	remaining: 8m 7s
1000:	test: 0.5019514	best: 0.5021630 (995)	total: 1m 8s	remaining: 8m
1100:	test: 0.5038158	best: 0.5039250 (1095)	total: 1m 15s	remaining: 7m 53s
1200:	test: 0.5047867	best: 0.5049844 (1180)	total: 1m 22s	remaining: 7m 45s
1300:	test: 0.5064481	best: 0.5064481 (130

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3537660	best: 0.3537660 (0)	total: 270ms	remaining: 35m 58s
100:	test: 0.4591970	best: 0.4591970 (100)	total: 20.4s	remaining: 26m 39s
200:	test: 0.4740540	best: 0.4740540 (200)	total: 40.5s	remaining: 26m 12s
300:	test: 0.4827673	best: 0.4827673 (300)	total: 1m	remaining: 25m 48s
400:	test: 0.4884890	best: 0.4884890 (400)	total: 1m 20s	remaining: 25m 23s
500:	test: 0.4938377	best: 0.4938377 (500)	total: 1m 40s	remaining: 25m 3s
600:	test: 0.4972008	best: 0.4972008 (600)	total: 2m	remaining: 24m 45s
700:	test: 0.4993689	best: 0.4993689 (700)	total: 2m 20s	remaining: 24m 22s
800:	test: 0.5019446	best: 0.5019446 (800)	total: 2m 40s	remaining: 24m 3s
900:	test: 0.5039389	best: 0.5039389 (900)	total: 3m	remaining: 23m 44s
1000:	test: 0.5055392	best: 0.5055392 (1000)	total: 3m 20s	remaining: 23m 23s
1100:	test: 0.5065588	best: 0.5066642 (1090)	total: 3m 40s	remaining: 23m 4s
1200:	test: 0.5075963	best: 0.5075963 (1200)	total: 4m	remaining: 22m 44s
1300:	test: 0.5081552	best: 0.50

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2843848	best: 0.2843848 (0)	total: 163ms	remaining: 21m 39s
100:	test: 0.4300956	best: 0.4300956 (100)	total: 2.06s	remaining: 2m 40s
200:	test: 0.4487852	best: 0.4487852 (200)	total: 3.99s	remaining: 2m 34s
300:	test: 0.4597920	best: 0.4597920 (300)	total: 5.92s	remaining: 2m 31s
400:	test: 0.4667776	best: 0.4667776 (400)	total: 7.76s	remaining: 2m 27s
500:	test: 0.4734644	best: 0.4734644 (500)	total: 9.6s	remaining: 2m 23s
600:	test: 0.4775790	best: 0.4776337 (585)	total: 11.4s	remaining: 2m 20s
700:	test: 0.4818710	best: 0.4818710 (700)	total: 13.3s	remaining: 2m 18s
800:	test: 0.4853845	best: 0.4853845 (800)	total: 15.2s	remaining: 2m 16s
900:	test: 0.4876149	best: 0.4876149 (900)	total: 17s	remaining: 2m 14s
1000:	test: 0.4895792	best: 0.4895792 (1000)	total: 18.9s	remaining: 2m 11s
1100:	test: 0.4914252	best: 0.4914252 (1100)	total: 20.7s	remaining: 2m 9s
1200:	test: 0.4931667	best: 0.4931667 (1200)	total: 22.6s	remaining: 2m 7s
1300:	test: 0.4942661	best: 0.4942661 (1

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3489517	best: 0.3489517 (0)	total: 132ms	remaining: 17m 37s
100:	test: 0.4574562	best: 0.4574562 (100)	total: 10.6s	remaining: 13m 45s
200:	test: 0.4710768	best: 0.4710768 (200)	total: 21s	remaining: 13m 34s
300:	test: 0.4799350	best: 0.4799350 (300)	total: 31.4s	remaining: 13m 23s
400:	test: 0.4861778	best: 0.4861778 (400)	total: 41.9s	remaining: 13m 13s
500:	test: 0.4913597	best: 0.4913941 (495)	total: 52.3s	remaining: 13m 3s
600:	test: 0.4954354	best: 0.4954354 (600)	total: 1m 2s	remaining: 12m 52s
700:	test: 0.4977140	best: 0.4977140 (700)	total: 1m 13s	remaining: 12m 42s
800:	test: 0.4991956	best: 0.4991956 (800)	total: 1m 23s	remaining: 12m 32s
900:	test: 0.5007129	best: 0.5009363 (885)	total: 1m 34s	remaining: 12m 22s
1000:	test: 0.5020065	best: 0.5021173 (950)	total: 1m 44s	remaining: 12m 11s
1100:	test: 0.5038896	best: 0.5038896 (1100)	total: 1m 55s	remaining: 12m 1s
1200:	test: 0.5051664	best: 0.5051664 (1200)	total: 2m 5s	remaining: 11m 50s
1300:	test: 0.5068500	b

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3584026	best: 0.3584026 (0)	total: 373ms	remaining: 49m 41s
100:	test: 0.4613619	best: 0.4613619 (100)	total: 35.4s	remaining: 46m 7s
200:	test: 0.4736045	best: 0.4736217 (195)	total: 1m 10s	remaining: 45m 27s
300:	test: 0.4820911	best: 0.4820911 (300)	total: 1m 45s	remaining: 44m 54s
400:	test: 0.4879874	best: 0.4879874 (400)	total: 2m 20s	remaining: 44m 20s
500:	test: 0.4927414	best: 0.4927414 (500)	total: 2m 55s	remaining: 43m 43s
600:	test: 0.4971898	best: 0.4971898 (600)	total: 3m 29s	remaining: 43m 4s
700:	test: 0.4990728	best: 0.4990728 (700)	total: 4m 4s	remaining: 42m 28s
800:	test: 0.5011627	best: 0.5011627 (800)	total: 4m 39s	remaining: 41m 55s
900:	test: 0.5031541	best: 0.5031541 (900)	total: 5m 14s	remaining: 41m 17s
1000:	test: 0.5048765	best: 0.5048765 (1000)	total: 5m 49s	remaining: 40m 44s
1100:	test: 0.5056916	best: 0.5060840 (1085)	total: 6m 24s	remaining: 40m 9s
1200:	test: 0.5072126	best: 0.5073507 (1195)	total: 6m 59s	remaining: 39m 32s
1300:	test: 0.50

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2832659	best: 0.2832659 (0)	total: 38.6ms	remaining: 5m 8s
100:	test: 0.4547397	best: 0.4547397 (100)	total: 1.57s	remaining: 2m 3s
200:	test: 0.4709152	best: 0.4709152 (200)	total: 3.13s	remaining: 2m 1s
300:	test: 0.4819358	best: 0.4821517 (295)	total: 4.71s	remaining: 2m
400:	test: 0.4879414	best: 0.4879414 (400)	total: 6.26s	remaining: 1m 58s
500:	test: 0.4928704	best: 0.4928704 (500)	total: 7.81s	remaining: 1m 56s
600:	test: 0.4949374	best: 0.4949374 (600)	total: 9.39s	remaining: 1m 55s
700:	test: 0.4974346	best: 0.4974346 (700)	total: 11s	remaining: 1m 54s
800:	test: 0.4985465	best: 0.4985465 (800)	total: 12.5s	remaining: 1m 52s
900:	test: 0.5000229	best: 0.5000229 (900)	total: 14.1s	remaining: 1m 51s
1000:	test: 0.5012959	best: 0.5012959 (1000)	total: 15.7s	remaining: 1m 49s
1100:	test: 0.5019060	best: 0.5021002 (1065)	total: 17.2s	remaining: 1m 48s
1200:	test: 0.5026624	best: 0.5026624 (1200)	total: 18.8s	remaining: 1m 46s
1300:	test: 0.5029672	best: 0.5030741 (1245)

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3454527	best: 0.3454527 (0)	total: 91.5ms	remaining: 12m 12s
100:	test: 0.4730514	best: 0.4730514 (100)	total: 6.93s	remaining: 9m 2s
200:	test: 0.4892124	best: 0.4892124 (200)	total: 13.8s	remaining: 8m 56s
300:	test: 0.4963297	best: 0.4963297 (300)	total: 20.8s	remaining: 8m 52s
400:	test: 0.5011501	best: 0.5014659 (395)	total: 27.8s	remaining: 8m 46s
500:	test: 0.5044926	best: 0.5044926 (500)	total: 34.6s	remaining: 8m 37s
600:	test: 0.5066097	best: 0.5066097 (600)	total: 41.4s	remaining: 8m 29s
700:	test: 0.5081985	best: 0.5081985 (700)	total: 48.2s	remaining: 8m 21s
800:	test: 0.5089746	best: 0.5089746 (800)	total: 55s	remaining: 8m 14s
900:	test: 0.5095413	best: 0.5096267 (895)	total: 1m 1s	remaining: 8m 6s
1000:	test: 0.5102031	best: 0.5104081 (990)	total: 1m 8s	remaining: 7m 59s
1100:	test: 0.5112569	best: 0.5114625 (1080)	total: 1m 15s	remaining: 7m 52s
1200:	test: 0.5114205	best: 0.5117779 (1170)	total: 1m 22s	remaining: 7m 44s
1300:	test: 0.5124858	best: 0.5124858

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3537660	best: 0.3537660 (0)	total: 226ms	remaining: 30m 7s
100:	test: 0.4754205	best: 0.4754205 (100)	total: 20.2s	remaining: 26m 18s
200:	test: 0.4902807	best: 0.4902807 (200)	total: 40s	remaining: 25m 52s
300:	test: 0.4985815	best: 0.4985815 (300)	total: 1m	remaining: 25m 38s
400:	test: 0.5028503	best: 0.5028503 (400)	total: 1m 20s	remaining: 25m 17s
500:	test: 0.5059195	best: 0.5060320 (490)	total: 1m 40s	remaining: 24m 59s
600:	test: 0.5083204	best: 0.5087246 (555)	total: 2m	remaining: 24m 38s
700:	test: 0.5103047	best: 0.5103047 (700)	total: 2m 19s	remaining: 24m 16s
800:	test: 0.5110774	best: 0.5114685 (785)	total: 2m 39s	remaining: 23m 57s
900:	test: 0.5113605	best: 0.5116695 (890)	total: 2m 59s	remaining: 23m 37s
1000:	test: 0.5124661	best: 0.5125086 (995)	total: 3m 20s	remaining: 23m 18s
1100:	test: 0.5126937	best: 0.5128031 (1025)	total: 3m 39s	remaining: 22m 57s
1200:	test: 0.5126596	best: 0.5128031 (1025)	total: 3m 59s	remaining: 22m 36s
1300:	test: 0.5125665	bes

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.2843848	best: 0.2843848 (0)	total: 37.4ms	remaining: 4m 59s
100:	test: 0.4537385	best: 0.4537385 (100)	total: 1.89s	remaining: 2m 27s
200:	test: 0.4711006	best: 0.4711006 (200)	total: 3.76s	remaining: 2m 26s
300:	test: 0.4818250	best: 0.4818872 (295)	total: 5.65s	remaining: 2m 24s
400:	test: 0.4875753	best: 0.4875753 (400)	total: 7.51s	remaining: 2m 22s
500:	test: 0.4927953	best: 0.4929054 (495)	total: 9.37s	remaining: 2m 20s
600:	test: 0.4957387	best: 0.4957387 (600)	total: 11.2s	remaining: 2m 18s
700:	test: 0.4982158	best: 0.4983876 (695)	total: 13.1s	remaining: 2m 16s
800:	test: 0.4997157	best: 0.4997157 (800)	total: 15s	remaining: 2m 14s
900:	test: 0.5006011	best: 0.5007593 (850)	total: 16.8s	remaining: 2m 12s
1000:	test: 0.5013133	best: 0.5014866 (955)	total: 18.7s	remaining: 2m 10s
1100:	test: 0.5020564	best: 0.5020564 (1100)	total: 20.6s	remaining: 2m 8s
1200:	test: 0.5028170	best: 0.5028598 (1180)	total: 22.4s	remaining: 2m 6s
1300:	test: 0.5034974	best: 0.5036085 (1

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3489517	best: 0.3489517 (0)	total: 137ms	remaining: 18m 13s
100:	test: 0.4742166	best: 0.4742166 (100)	total: 10.6s	remaining: 13m 52s
200:	test: 0.4892077	best: 0.4894336 (195)	total: 21.2s	remaining: 13m 40s
300:	test: 0.4978101	best: 0.4978101 (300)	total: 31.6s	remaining: 13m 29s
400:	test: 0.5018169	best: 0.5018169 (400)	total: 42.2s	remaining: 13m 19s
500:	test: 0.5044247	best: 0.5044247 (500)	total: 53s	remaining: 13m 13s
600:	test: 0.5064760	best: 0.5064760 (600)	total: 1m 3s	remaining: 13m 6s
700:	test: 0.5078261	best: 0.5078261 (700)	total: 1m 14s	remaining: 12m 57s
800:	test: 0.5096342	best: 0.5096342 (800)	total: 1m 25s	remaining: 12m 47s
900:	test: 0.5095560	best: 0.5102379 (875)	total: 1m 35s	remaining: 12m 36s
1000:	test: 0.5094319	best: 0.5102379 (875)	total: 1m 46s	remaining: 12m 24s
1100:	test: 0.5109188	best: 0.5109188 (1100)	total: 1m 56s	remaining: 12m 12s
1200:	test: 0.5114473	best: 0.5118990 (1185)	total: 2m 7s	remaining: 12m
1300:	test: 0.5121476	best

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3584026	best: 0.3584026 (0)	total: 372ms	remaining: 49m 34s
100:	test: 0.4762367	best: 0.4762367 (100)	total: 35.4s	remaining: 46m 6s
200:	test: 0.4920495	best: 0.4920495 (200)	total: 1m 10s	remaining: 45m 31s
300:	test: 0.4997464	best: 0.4997464 (300)	total: 1m 45s	remaining: 45m 1s
400:	test: 0.5030496	best: 0.5030496 (400)	total: 2m 20s	remaining: 44m 25s
500:	test: 0.5053317	best: 0.5056810 (495)	total: 2m 55s	remaining: 43m 44s
600:	test: 0.5077339	best: 0.5078442 (590)	total: 3m 30s	remaining: 43m 9s
700:	test: 0.5091636	best: 0.5091636 (700)	total: 4m 5s	remaining: 42m 37s
800:	test: 0.5099381	best: 0.5099381 (800)	total: 4m 40s	remaining: 41m 57s
900:	test: 0.5112712	best: 0.5113774 (895)	total: 5m 14s	remaining: 41m 21s
1000:	test: 0.5119653	best: 0.5121908 (955)	total: 5m 49s	remaining: 40m 46s
1100:	test: 0.5123929	best: 0.5128264 (1055)	total: 6m 24s	remaining: 40m 11s
1200:	test: 0.5138055	best: 0.5139203 (1195)	total: 7m	remaining: 39m 38s
1300:	test: 0.5145888

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3396586	best: 0.3396586 (0)	total: 25.7ms	remaining: 3m 25s
100:	test: 0.4645902	best: 0.4645902 (100)	total: 1.14s	remaining: 1m 28s
200:	test: 0.4818895	best: 0.4818895 (200)	total: 2.25s	remaining: 1m 27s
300:	test: 0.4917228	best: 0.4917228 (300)	total: 3.35s	remaining: 1m 25s
400:	test: 0.4964158	best: 0.4964158 (400)	total: 4.45s	remaining: 1m 24s
500:	test: 0.4994928	best: 0.4994928 (500)	total: 5.55s	remaining: 1m 23s
600:	test: 0.5006095	best: 0.5006095 (600)	total: 6.66s	remaining: 1m 22s
700:	test: 0.5028708	best: 0.5030458 (680)	total: 7.77s	remaining: 1m 20s
800:	test: 0.5044033	best: 0.5046584 (790)	total: 8.88s	remaining: 1m 19s
900:	test: 0.5057920	best: 0.5057920 (900)	total: 9.97s	remaining: 1m 18s
1000:	test: 0.5062954	best: 0.5062954 (1000)	total: 11.1s	remaining: 1m 17s
1100:	test: 0.5075751	best: 0.5075751 (1100)	total: 12.2s	remaining: 1m 16s
1200:	test: 0.5085502	best: 0.5085502 (1200)	total: 13.3s	remaining: 1m 15s
1300:	test: 0.5095973	best: 0.50959

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3808221	best: 0.3808221 (0)	total: 84.2ms	remaining: 11m 13s
100:	test: 0.4875489	best: 0.4875489 (100)	total: 6.51s	remaining: 8m 28s
200:	test: 0.5005493	best: 0.5005493 (200)	total: 12.9s	remaining: 8m 21s
300:	test: 0.5043690	best: 0.5046690 (295)	total: 19.3s	remaining: 8m 13s
400:	test: 0.5073735	best: 0.5073735 (400)	total: 25.7s	remaining: 8m 6s
500:	test: 0.5097180	best: 0.5097180 (500)	total: 32s	remaining: 7m 59s
600:	test: 0.5121403	best: 0.5121403 (600)	total: 38.5s	remaining: 7m 53s
700:	test: 0.5134669	best: 0.5134669 (700)	total: 44.9s	remaining: 7m 47s
800:	test: 0.5139930	best: 0.5139930 (800)	total: 51.3s	remaining: 7m 41s
900:	test: 0.5149078	best: 0.5152532 (890)	total: 57.7s	remaining: 7m 34s
1000:	test: 0.5144914	best: 0.5155797 (915)	total: 1m 4s	remaining: 7m 29s
1100:	test: 0.5151849	best: 0.5155797 (915)	total: 1m 10s	remaining: 7m 22s
1200:	test: 0.5153664	best: 0.5157104 (1190)	total: 1m 17s	remaining: 7m 16s
1300:	test: 0.5154186	best: 0.5159262

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3937491	best: 0.3937491 (0)	total: 213ms	remaining: 28m 26s
100:	test: 0.4899916	best: 0.4899916 (100)	total: 19.9s	remaining: 25m 52s
200:	test: 0.5021621	best: 0.5021621 (200)	total: 39.5s	remaining: 25m 34s
300:	test: 0.5083251	best: 0.5083251 (300)	total: 59.2s	remaining: 25m 13s
400:	test: 0.5109898	best: 0.5110151 (395)	total: 1m 18s	remaining: 24m 54s
500:	test: 0.5128683	best: 0.5130768 (485)	total: 1m 38s	remaining: 24m 32s
600:	test: 0.5141400	best: 0.5144418 (590)	total: 1m 58s	remaining: 24m 13s
700:	test: 0.5151847	best: 0.5151847 (700)	total: 2m 17s	remaining: 23m 53s
800:	test: 0.5158273	best: 0.5158273 (800)	total: 2m 37s	remaining: 23m 33s
900:	test: 0.5156279	best: 0.5160189 (835)	total: 2m 57s	remaining: 23m 14s
1000:	test: 0.5160764	best: 0.5160764 (1000)	total: 3m 16s	remaining: 22m 55s
1100:	test: 0.5165435	best: 0.5166111 (1095)	total: 3m 36s	remaining: 22m 35s
1200:	test: 0.5169462	best: 0.5169462 (1200)	total: 3m 55s	remaining: 22m 15s
1300:	test: 0.

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3079449	best: 0.3079449 (0)	total: 27.9ms	remaining: 3m 43s
100:	test: 0.4666365	best: 0.4666365 (100)	total: 1.45s	remaining: 1m 53s
200:	test: 0.4820682	best: 0.4820682 (200)	total: 2.87s	remaining: 1m 51s
300:	test: 0.4916437	best: 0.4916437 (300)	total: 4.31s	remaining: 1m 50s
400:	test: 0.4956863	best: 0.4956863 (400)	total: 5.73s	remaining: 1m 48s
500:	test: 0.4985395	best: 0.4985395 (500)	total: 7.13s	remaining: 1m 46s
600:	test: 0.5007200	best: 0.5007200 (600)	total: 8.55s	remaining: 1m 45s
700:	test: 0.5024944	best: 0.5026268 (685)	total: 9.96s	remaining: 1m 43s
800:	test: 0.5041915	best: 0.5042236 (790)	total: 11.4s	remaining: 1m 42s
900:	test: 0.5050491	best: 0.5050679 (860)	total: 12.8s	remaining: 1m 40s
1000:	test: 0.5057295	best: 0.5057295 (1000)	total: 14.2s	remaining: 1m 39s
1100:	test: 0.5072951	best: 0.5072951 (1100)	total: 15.6s	remaining: 1m 38s
1200:	test: 0.5077672	best: 0.5079164 (1185)	total: 17.1s	remaining: 1m 36s
1300:	test: 0.5090205	best: 0.50905

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3813925	best: 0.3813925 (0)	total: 114ms	remaining: 15m 10s
100:	test: 0.4874804	best: 0.4874804 (100)	total: 10.1s	remaining: 13m 13s
200:	test: 0.5005671	best: 0.5005671 (200)	total: 20.2s	remaining: 13m 3s
300:	test: 0.5051897	best: 0.5051897 (300)	total: 30.2s	remaining: 12m 53s
400:	test: 0.5082366	best: 0.5082366 (400)	total: 40.3s	remaining: 12m 43s
500:	test: 0.5108816	best: 0.5110181 (495)	total: 50.3s	remaining: 12m 32s
600:	test: 0.5121859	best: 0.5121859 (600)	total: 1m	remaining: 12m 22s
700:	test: 0.5123716	best: 0.5126230 (695)	total: 1m 10s	remaining: 12m 12s
800:	test: 0.5135917	best: 0.5141351 (780)	total: 1m 20s	remaining: 12m 2s
900:	test: 0.5142892	best: 0.5146070 (895)	total: 1m 30s	remaining: 11m 52s
1000:	test: 0.5149336	best: 0.5151466 (995)	total: 1m 40s	remaining: 11m 42s
1100:	test: 0.5160050	best: 0.5160050 (1100)	total: 1m 50s	remaining: 11m 32s
1200:	test: 0.5163446	best: 0.5163446 (1200)	total: 2m	remaining: 11m 22s
1300:	test: 0.5162711	best:

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3908396	best: 0.3908396 (0)	total: 443ms	remaining: 59m 6s
100:	test: 0.4902641	best: 0.4902641 (100)	total: 35.2s	remaining: 45m 52s
200:	test: 0.5030439	best: 0.5030439 (200)	total: 1m 9s	remaining: 45m 12s
300:	test: 0.5065112	best: 0.5065112 (300)	total: 1m 44s	remaining: 44m 35s
400:	test: 0.5099559	best: 0.5099559 (400)	total: 2m 19s	remaining: 43m 59s
500:	test: 0.5109514	best: 0.5109514 (500)	total: 2m 53s	remaining: 43m 24s
600:	test: 0.5124313	best: 0.5124313 (600)	total: 3m 28s	remaining: 42m 50s
700:	test: 0.5141376	best: 0.5141785 (685)	total: 4m 3s	remaining: 42m 15s
800:	test: 0.5149300	best: 0.5149300 (800)	total: 4m 38s	remaining: 41m 40s
900:	test: 0.5155613	best: 0.5155978 (895)	total: 5m 12s	remaining: 41m 5s
1000:	test: 0.5164951	best: 0.5168061 (990)	total: 5m 47s	remaining: 40m 30s
1100:	test: 0.5170265	best: 0.5176748 (1080)	total: 6m 22s	remaining: 39m 55s
1200:	test: 0.5171455	best: 0.5178558 (1145)	total: 6m 57s	remaining: 39m 21s
1300:	test: 0.517

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3396586	best: 0.3396586 (0)	total: 24.6ms	remaining: 3m 17s
100:	test: 0.4858359	best: 0.4858359 (100)	total: 1.17s	remaining: 1m 31s
200:	test: 0.4968650	best: 0.4969942 (195)	total: 2.29s	remaining: 1m 28s
300:	test: 0.5022683	best: 0.5022683 (300)	total: 3.42s	remaining: 1m 27s
400:	test: 0.5045310	best: 0.5045310 (400)	total: 4.55s	remaining: 1m 26s
500:	test: 0.5065694	best: 0.5065694 (500)	total: 5.68s	remaining: 1m 24s
600:	test: 0.5082050	best: 0.5084063 (595)	total: 6.81s	remaining: 1m 23s
700:	test: 0.5095120	best: 0.5095120 (700)	total: 7.94s	remaining: 1m 22s
800:	test: 0.5110234	best: 0.5110234 (800)	total: 9.07s	remaining: 1m 21s
900:	test: 0.5114932	best: 0.5114932 (900)	total: 10.2s	remaining: 1m 20s
1000:	test: 0.5113750	best: 0.5117218 (950)	total: 11.3s	remaining: 1m 19s
1100:	test: 0.5121100	best: 0.5121100 (1100)	total: 12.5s	remaining: 1m 18s
1200:	test: 0.5115446	best: 0.5123655 (1105)	total: 13.6s	remaining: 1m 16s
1300:	test: 0.5121277	best: 0.512365

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3808221	best: 0.3808221 (0)	total: 81.1ms	remaining: 10m 48s
100:	test: 0.5011994	best: 0.5011994 (100)	total: 6.44s	remaining: 8m 23s
200:	test: 0.5094493	best: 0.5094493 (200)	total: 12.8s	remaining: 8m 15s
300:	test: 0.5122579	best: 0.5122579 (300)	total: 19.1s	remaining: 8m 9s
400:	test: 0.5141866	best: 0.5143890 (395)	total: 25.5s	remaining: 8m 2s
500:	test: 0.5150082	best: 0.5152459 (485)	total: 31.8s	remaining: 7m 56s
600:	test: 0.5156084	best: 0.5156182 (545)	total: 38.2s	remaining: 7m 49s
700:	test: 0.5144447	best: 0.5158662 (615)	total: 44.5s	remaining: 7m 43s
800:	test: 0.5149595	best: 0.5158662 (615)	total: 50.8s	remaining: 7m 36s
900:	test: 0.5158264	best: 0.5159993 (880)	total: 57.1s	remaining: 7m 30s
1000:	test: 0.5148530	best: 0.5161167 (905)	total: 1m 3s	remaining: 7m 23s
1100:	test: 0.5140584	best: 0.5161167 (905)	total: 1m 9s	remaining: 7m 17s
1200:	test: 0.5137147	best: 0.5161167 (905)	total: 1m 16s	remaining: 7m 11s
1300:	test: 0.5139835	best: 0.5161167 

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3937491	best: 0.3937491 (0)	total: 220ms	remaining: 29m 21s
100:	test: 0.5031382	best: 0.5031382 (100)	total: 19.9s	remaining: 25m 53s
200:	test: 0.5097076	best: 0.5097076 (200)	total: 39.5s	remaining: 25m 31s
300:	test: 0.5126055	best: 0.5129730 (280)	total: 59.1s	remaining: 25m 10s
400:	test: 0.5136146	best: 0.5136146 (400)	total: 1m 18s	remaining: 24m 51s
500:	test: 0.5140807	best: 0.5143157 (450)	total: 1m 38s	remaining: 24m 31s
600:	test: 0.5129134	best: 0.5144548 (565)	total: 1m 58s	remaining: 24m 12s
700:	test: 0.5131502	best: 0.5144548 (565)	total: 2m 17s	remaining: 23m 53s
800:	test: 0.5125719	best: 0.5144548 (565)	total: 2m 37s	remaining: 23m 33s
900:	test: 0.5120587	best: 0.5144548 (565)	total: 2m 56s	remaining: 23m 14s
1000:	test: 0.5108630	best: 0.5144548 (565)	total: 3m 16s	remaining: 22m 54s
1100:	test: 0.5119819	best: 0.5144548 (565)	total: 3m 36s	remaining: 22m 35s
1200:	test: 0.5113504	best: 0.5144548 (565)	total: 3m 56s	remaining: 22m 16s
1300:	test: 0.510

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3079449	best: 0.3079449 (0)	total: 46.2ms	remaining: 6m 9s
100:	test: 0.4863003	best: 0.4863003 (100)	total: 1.49s	remaining: 1m 56s
200:	test: 0.4979748	best: 0.4979748 (200)	total: 2.92s	remaining: 1m 53s
300:	test: 0.5028801	best: 0.5028801 (300)	total: 4.35s	remaining: 1m 51s
400:	test: 0.5056845	best: 0.5056845 (400)	total: 5.78s	remaining: 1m 49s
500:	test: 0.5085069	best: 0.5085069 (500)	total: 7.21s	remaining: 1m 47s
600:	test: 0.5099661	best: 0.5099661 (600)	total: 8.65s	remaining: 1m 46s
700:	test: 0.5103576	best: 0.5110568 (660)	total: 10.1s	remaining: 1m 44s
800:	test: 0.5114725	best: 0.5114725 (800)	total: 11.5s	remaining: 1m 43s
900:	test: 0.5126492	best: 0.5126492 (900)	total: 12.9s	remaining: 1m 41s
1000:	test: 0.5128168	best: 0.5129162 (925)	total: 14.3s	remaining: 1m 40s
1100:	test: 0.5136212	best: 0.5137992 (1095)	total: 15.8s	remaining: 1m 38s
1200:	test: 0.5142587	best: 0.5144612 (1190)	total: 17.2s	remaining: 1m 37s
1300:	test: 0.5139490	best: 0.5144612

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3813925	best: 0.3813925 (0)	total: 145ms	remaining: 19m 21s
100:	test: 0.5010819	best: 0.5010819 (100)	total: 10.2s	remaining: 13m 14s
200:	test: 0.5074701	best: 0.5074701 (200)	total: 20.2s	remaining: 13m 2s
300:	test: 0.5122345	best: 0.5122345 (300)	total: 30.2s	remaining: 12m 51s
400:	test: 0.5134700	best: 0.5135739 (390)	total: 40.2s	remaining: 12m 41s
500:	test: 0.5145482	best: 0.5150688 (420)	total: 50.1s	remaining: 12m 30s
600:	test: 0.5142797	best: 0.5150688 (420)	total: 1m	remaining: 12m 20s
700:	test: 0.5153263	best: 0.5157414 (690)	total: 1m 10s	remaining: 12m 10s
800:	test: 0.5141184	best: 0.5157414 (690)	total: 1m 20s	remaining: 11m 59s
900:	test: 0.5149766	best: 0.5157414 (690)	total: 1m 30s	remaining: 11m 50s
1000:	test: 0.5141950	best: 0.5157414 (690)	total: 1m 40s	remaining: 11m 40s
1100:	test: 0.5139647	best: 0.5157414 (690)	total: 1m 50s	remaining: 11m 30s
1200:	test: 0.5137311	best: 0.5157414 (690)	total: 2m	remaining: 11m 20s
1300:	test: 0.5132668	best: 

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3908396	best: 0.3908396 (0)	total: 407ms	remaining: 54m 19s
100:	test: 0.5045735	best: 0.5045735 (100)	total: 35.1s	remaining: 45m 45s
200:	test: 0.5106915	best: 0.5106915 (200)	total: 1m 9s	remaining: 45m 9s
300:	test: 0.5152330	best: 0.5153584 (295)	total: 1m 44s	remaining: 44m 33s
400:	test: 0.5148759	best: 0.5153584 (295)	total: 2m 19s	remaining: 43m 58s
500:	test: 0.5151853	best: 0.5156784 (475)	total: 2m 53s	remaining: 43m 23s
600:	test: 0.5153569	best: 0.5158866 (560)	total: 3m 28s	remaining: 42m 48s
700:	test: 0.5151275	best: 0.5158866 (560)	total: 4m 3s	remaining: 42m 13s
800:	test: 0.5153886	best: 0.5158866 (560)	total: 4m 38s	remaining: 41m 38s
900:	test: 0.5159860	best: 0.5164095 (880)	total: 5m 12s	remaining: 41m 4s
1000:	test: 0.5153562	best: 0.5164095 (880)	total: 5m 47s	remaining: 40m 29s
1100:	test: 0.5145624	best: 0.5164095 (880)	total: 6m 22s	remaining: 39m 55s
1200:	test: 0.5141576	best: 0.5164095 (880)	total: 6m 57s	remaining: 39m 20s
1300:	test: 0.51316

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3808221	best: 0.3808221 (0)	total: 185ms	remaining: 24m 36s
100:	test: 0.4875489	best: 0.4875489 (100)	total: 6.63s	remaining: 8m 38s
200:	test: 0.5005493	best: 0.5005493 (200)	total: 13s	remaining: 8m 25s
300:	test: 0.5043690	best: 0.5046690 (295)	total: 19.4s	remaining: 8m 16s
400:	test: 0.5073735	best: 0.5073735 (400)	total: 25.8s	remaining: 8m 9s
500:	test: 0.5097180	best: 0.5097180 (500)	total: 32.2s	remaining: 8m 2s
600:	test: 0.5121403	best: 0.5121403 (600)	total: 38.7s	remaining: 7m 56s
700:	test: 0.5134669	best: 0.5134669 (700)	total: 45.1s	remaining: 7m 49s
800:	test: 0.5139930	best: 0.5139930 (800)	total: 51.5s	remaining: 7m 43s
900:	test: 0.5149078	best: 0.5152532 (890)	total: 57.9s	remaining: 7m 36s
1000:	test: 0.5144914	best: 0.5155797 (915)	total: 1m 4s	remaining: 7m 29s
1100:	test: 0.5151849	best: 0.5155797 (915)	total: 1m 10s	remaining: 7m 23s
1200:	test: 0.5153664	best: 0.5157104 (1190)	total: 1m 17s	remaining: 7m 16s
1300:	test: 0.5154186	best: 0.5159262 (

In [ ]:
for seed in [1337]:
    params = {'loss_function': 'QueryRMSE', 'random_seed': seed, 'iterations': 7500, 'learning_rate': 0.01, 'max_bin': 128, 'depth': 10, 'task_type': 'GPU', 'eval_metric': 'NDCG:top=10;type=Exp'}
    model = Model()
    model.fit(df_train, df_valid, params=params)

Default metric period is 5 because NDCG is/are not implemented for GPU
Metric NDCG:top=10;type=Exp is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time


0:	test: 0.3843786	best: 0.3843786 (0)	total: 81.2ms	remaining: 10m 9s
100:	test: 0.4649174	best: 0.4649174 (100)	total: 6.54s	remaining: 7m 58s
200:	test: 0.4783491	best: 0.4783491 (200)	total: 13s	remaining: 7m 50s
300:	test: 0.4877500	best: 0.4879703 (295)	total: 19.4s	remaining: 7m 44s
400:	test: 0.4947765	best: 0.4947765 (400)	total: 25.8s	remaining: 7m 37s
500:	test: 0.4984680	best: 0.4984680 (500)	total: 32.2s	remaining: 7m 29s
600:	test: 0.5004852	best: 0.5004852 (600)	total: 38.6s	remaining: 7m 22s
700:	test: 0.5034260	best: 0.5036388 (685)	total: 45s	remaining: 7m 16s
800:	test: 0.5043121	best: 0.5044907 (770)	total: 51.4s	remaining: 7m 10s
900:	test: 0.5058795	best: 0.5058795 (900)	total: 57.8s	remaining: 7m 3s
1000:	test: 0.5074456	best: 0.5074456 (1000)	total: 1m 4s	remaining: 6m 57s
1100:	test: 0.5083134	best: 0.5084515 (1085)	total: 1m 10s	remaining: 6m 50s
1200:	test: 0.5092884	best: 0.5094781 (1195)	total: 1m 17s	remaining: 6m 44s
1300:	test: 0.5095813	best: 0.5098989 

Теперь, имея предикты, можно посчитать метрики качества:

In [41]:
def compute_metrics(y_true, y_hat, q):
    # List of metrics to evaluate
    eval_metrics = ['NDCG:top=10;type=Exp']
    
    for eval_metric in eval_metrics:
        scores = utils.eval_metric(y_true, y_hat, eval_metric, group_id=q)
    
        # Print scores
        print(f"metric = {eval_metric} score = {scores[0]:.3f}")

y_hat_test = model.predict(df_test)

# Get test targets and groups
y_test = df_test['label'].to_numpy()
q_test = df_test['qid'].to_numpy().astype('uint32')
    
# Compute metrics on test
compute_metrics(y_test, y_hat_test, q_test)

metric = NDCG:top=10;type=Exp score = 0.511


Ожидаем, что ваша модель покажет результаты выше бейзлайна!

In [42]:
model.model.save_model("final.cbm")

In [49]:
!unzip model.zip

Archive:  model.zip
  inflating: final.cbm               


In [50]:
model = Model()
model.model = CatBoostRanker().load_model("final.cbm")

In [51]:
def compute_metrics(y_true, y_hat, q):
    # List of metrics to evaluate
    eval_metrics = ['NDCG:top=10;type=Exp']
    
    for eval_metric in eval_metrics:
        scores = utils.eval_metric(y_true, y_hat, eval_metric, group_id=q)
    
        # Print scores
        print(f"metric = {eval_metric} score = {scores[0]:.3f}")

y_hat_test = model.predict(df_test)

# Get test targets and groups
y_test = df_test['label'].to_numpy()
q_test = df_test['qid'].to_numpy().astype('uint32')
    
# Compute metrics on test
compute_metrics(y_test, y_hat_test, q_test)

metric = NDCG:top=10;type=Exp score = 0.511
